### PR TITLE
feat(hierarchy-list): indeterminate checkbox handling to items in MultipleNesting mode.

### LINKED ISSUE #3004: [HierarchyList] No indeterminate state on nested hierarchy checkbox when in multiple mode
### What package is this for?

- [X] React
- [ ] Angular

### Describe the bug

No handling of indeterminate state on HierarchyList in multiple-nesting edit style.
### To Reproduce

1. Go to https://next.carbon-addons-iot-react.com/?path=/story/1-watson-iot-hierarchylist--with-nested-reorder
2. Change "Editing Style" knob to "multiple-nesting"
3. Check checkbox on "Atlanta Braves"
4. Click chevron to open "Atlanta Braves"
5. Uncheck "Dansby Swanson"
6. "Atlanta Braves" should now show the indeterminate checkbox state.

### PR DESCRIPTION
**Pull request by kevinsperrine**
Closes #3004 

### Depends on #3064 and #3065

**Summary**

- Add indeterminate state handling to HierarchyList in MultipleNesting mode

**Change List (commits, features, bugs, etc)**

- add indeterminateIds state and pass that to the list
- adjust the `handleEditModeSelect` function to mark the parent as selected if all children are.
- add useEffect to HierarchyList to keep indeterminateIds updated as selected ids are changed.
- add tests to confirm

**Acceptance Test (how to verify the PR)**

1. https://deploy-preview-3066--carbon-addons-iot-react.netlify.app/?path=/story/1-watson-iot-hierarchylist--with-nested-reorder
2. turn on MultipleNesting mode
3. confirm New York Mets is indeterminate
4. Check the New York Mets checkbox
5. confirm New York Mets is still indeterminate (b/c of the locked row)
6. Check the Atlanta Braves
7. Uncheck an individual player
8. Confirm the Atlanta Braves is indeterminate
9. Re-check the player
10. 8. Confirm the Atlanta Braves is checked
11. repeat with isVirtualList

**Regression Test (how to make sure this PR doesn't break old functionality)**

- tests here

**Things to look for during review**

- [ ] Make sure all references to `iot` or `bx` class prefix is using the prefix variable
- [ ] (React) All major areas have a `data-testid` attribute. New test ids should have test written to ensure they are not changed or removed.
- [ ] UI should be checked in RTL mode to ensure the proper handling of layout and text.
- [ ] All strings should be translatable.
- [ ] The code should pass a11y scans (The storybook a11y knob should show no violations). New components should have a11y test written.
- [ ] Unit test should be written and should have a coverage of 90% or higher in all areas.
- [ ] All components should be passing visual regression test. For new styles or components either a visual regression test should be written for all permutations or the base image updated.
- [ ] Changes or new components should either write new or update existing documentation.
- [ ] PR should link and close out an existing issue

### CODE REVIEW

File: packages/react/src/components/List/HierarchyList/HierarchyList.test.jsx [hunk 1]
```diff
@@ -830,32 +830,101 @@ describe('HierarchyList', () => {
...
+        editingStyle={EditingStyle.MultipleNesting}
+        defaultExpandedIds={['Atlanta Braves']}
+        defaultSelectedId="Atlanta Braves_Ronald Acuna Jr."
+        title="indeterminate Ids"
+      />
+    );
+
+    expect(screen.getByTestId('Atlanta Braves-checkbox')).toBePartiallyChecked();
+    userEvent.click(screen.getByTestId('Atlanta Braves_Ronald Acuna Jr.-checkbox'));
+    expect(screen.getByTestId('Atlanta Braves-checkbox')).not.toBeChecked();
```
Comment (bjornalm) [inline packages/react/src/components/List/HierarchyList/HierarchyList.test.jsx:904]: not.toBePartiallyChecked()?

File: packages/react/src/components/List/HierarchyList/HierarchyList.test.jsx [hunk 2]
```diff
@@ -830,32 +830,101 @@ describe('HierarchyList', () => {
...
-      expect(screen.getByText('Page 2')).toBeVisible();
-      expect(screen.getByText('Item 9')).toBeVisible();
-    });
+    userEvent.click(screen.getByTestId('Atlanta Braves_Ronald Acuna Jr.-checkbox'));
+    expect(screen.getByTestId('Atlanta Braves-checkbox')).toBePartiallyChecked();
+    userEvent.click(screen.getByTestId('Atlanta Braves_Ronald Acuna Jr.-checkbox'));
+    expect(screen.getByTestId('Atlanta Braves-checkbox')).not.toBeChecked();
+  });
+
+  it('should unset indeterminate state on parent when a child is unchecked', () => {
```
Comment (bjornalm) [inline packages/react/src/components/List/HierarchyList/HierarchyList.test.jsx:891]: "should unset indeterminate state on parent when a PRESELECTED child is unchecked"
just testing the uncheck of a child is already tested in the test above

File: packages/react/src/components/List/HierarchyList/HierarchyList.test.jsx [hunk 3]
```diff
@@ -830,32 +830,101 @@ describe('HierarchyList', () => {
...
+      />
+    );
 
-      expect(screen.getByText('Page 2')).toBeVisible();
-      expect(screen.getByText('Item 9')).toBeVisible();
-    });
+    userEvent.click(screen.getByTestId('Atlanta Braves_Ronald Acuna Jr.-checkbox'));
+    expect(screen.getByTestId('Atlanta Braves-checkbox')).toBePartiallyChecked();
+    userEvent.click(screen.getByTestId('Atlanta Braves_Ronald Acuna Jr.-checkbox'));
+    expect(screen.getByTestId('Atlanta Braves-checkbox')).not.toBeChecked();
```
Comment (bjornalm) [inline packages/react/src/components/List/HierarchyList/HierarchyList.test.jsx:888]: not.toBePartiallyChecked()?

**Review by bjornalm — APPROVED** [review]
Code looks good, just a couple of comments on the tests and I would like to test it before approval but the preview build has failed.

**Review by bjornalm — CHANGES_REQUESTED** [review]
build preview build failed

File: packages/react/src/components/List/HierarchyList/HierarchyList.test.jsx (hunk 3 above)
Comment (kevinsperrine) [reply packages/react/src/components/List/HierarchyList/HierarchyList.test.jsx:888]: added

File: packages/react/src/components/List/HierarchyList/HierarchyList.test.jsx (hunk 1 above)
Comment (kevinsperrine) [reply packages/react/src/components/List/HierarchyList/HierarchyList.test.jsx:905]: added

**Review by davidicus — APPROVED** [review]

File: packages/react/src/components/List/HierarchyList/HierarchyList.test.jsx (hunk 2 above)
Comment (kevinsperrine) [reply packages/react/src/components/List/HierarchyList/HierarchyList.test.jsx]: renamed

**Review by bjornalm — APPROVED** [review]

**Pull request merged**